module-apps/apps-common/widgets/DateSetSpinner.cpp

```diff
@@ -0,0 +1,267 @@
+        : HBox(parent, x, y, w, h)
+    {
+        setAlignment(Alignment(Alignment::Horizontal::Center, Alignment::Vertical::Center));
+        setEdges(RectangleEdge::None);
+
+        title = new TextFixedSize(titleBox);
+        title->setMaximumSize(style::bell_base_layout::w, style::bell_base_layout::outer_layouts_h);
+        title->setFont(style::bell_sidelist_item::title_font);
+        title->setEdges(gui::RectangleEdge::None);
+        title->activeItem = false;
+        title->setAlignment(Alignment(gui::Alignment::Horizontal::Center, gui::Alignment::Vertical::Center));
+        title->setRichText(utils::translate("app_settings_title_day"));
+        title->drawUnderline(false);
+
+        attachDateField(day, U8IntegerSpinnerFixed::range{dayMin, dayMax, step});
+        attachSlash(firstSlash);
+        attachDateField(month, U8IntegerSpinnerFixed::range{monthMin, monthMax, step});
+        attachSlash(secondSlash);
+        attachDateField(year,
+                        U16IntegerSpinnerFixed::range{utils::time::Locale::min_years,
+                                                      utils::time::Locale::max_years,
+                                                      static_cast<std::uint16_t>(step)});
+
+        resizeItems();
+        setFont();
+
+        focusChangedCallback = [&](Item &) {
+            if (editMode != EditMode::Edit) {
+                setFocusItem(nullptr);
+                return false;
+            }
+
+            if (lastFocus != nullptr) {
+                updateFocus(lastFocus);
+            }
+            else {
+                updateFocus(day);
+            }
+
+            return true;
+        };
+        if (editMode == EditMode::Edit) {
+            updateFocus(day);
+        }
+    }
+
+    date::year_month_day DateSetSpinner::getDate()
+    {
+        return date::year(year->get_value()) / date::month(month->get_value()) / date::day(day->get_value());
+    }
+
+    void DateSetSpinner::setDate(const date::year_month_day date)
+    {
+        day->set_value(static_cast<unsigned>(date.day()));
+        month->set_value(static_cast<unsigned>(date.month()));
+        year->set_value(static_cast<int>(date.year()));
+        applySizeRestrictions();
+    }
+
+    void DateSetSpinner::applySizeRestrictions()
+    {
+        day->setMinimumWidthToFitText();
+        firstSlash->setMinimumWidthToFitText();
+        month->setMinimumWidthToFitText();
+        secondSlash->setMinimumWidthToFitText();
+        year->setMinimumWidthToFitText();
+
+        setMinimumSize(getWidgetMinimumAreaWidth(), getFontHeight(noFocusFontName));
+        setMaximumWidth(widgetMaximumArea.w);
+
+        HBox::informContentChanged();
+    }
+
+    std::uint32_t DateSetSpinner::getWidgetMinimumAreaWidth()
+    {
+        return day->widgetMinimumArea.w + firstSlash->widgetMinimumArea.w + month->widgetMinimumArea.w +
+               secondSlash->widgetMinimumArea.w + year->widgetMinimumArea.w;
```

Comment:
are those slashes different in width?

Reply:
no, would be better for you firstSlash->widgetMinimumArea.w * 2  ?

Reply:
I believe it's good enough solution.